backend/src/main/java/org/cryptomator/hub/RemoteUserPuller.java

```diff
@@ -30,17 +30,38 @@ void sync() {
 	void sync(Map<String, Group> keycloakGroups, Map<String, User> keycloakUsers) {
 		var databaseGroups = Group.<Group>findAll().stream().collect(Collectors.toMap(g -> g.id, Function.identity()));
 		var databaseUsers = User.<User>findAll().stream().collect(Collectors.toMap(u -> u.id, Function.identity()));
+		sync(keycloakGroups, keycloakUsers, databaseGroups, databaseUsers);
+	}
+
+	//visible for testing
+	void sync(Map<String, Group> keycloakGroups, Map<String, User> keycloakUsers, Map<String, Group> databaseGroups, Map<String, User> databaseUsers) {
+		syncAddedUsers(keycloakUsers, databaseUsers);
+		var deletedUsers = syncDeletedUsers(keycloakUsers, databaseUsers);
+		syncUpdatedUsers(keycloakUsers, databaseUsers, deletedUsers);
+		syncAddedGroups(keycloakGroups, databaseGroups);
+		var deletedGroups = syncDeletedGroups(keycloakGroups, databaseGroups);
+		syncUpdatedGroups(keycloakGroups, databaseGroups, deletedGroups);
```

Comment:
would generics help to avoid duplication of these three methods?

Reply:
Yes, a lot :heart_eyes:, implemented in https://github.com/cryptomator/hub/pull/105/commits/e6c560df17cd38797ecb6c2d3ba7330b950cfb13